pkg/scheduler/plugins/elasticquota/core/quota_tree_wrapper.go

```diff
@@ -0,0 +1,481 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"fmt"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+)
+
+type quotaNode struct {
+	quotaName         QuotaNameType
+	request           int64
+	SharedWeight      int64
+	minQuota          int64
+	runtimeQuota      int64
+	allowLentResource bool
+}
+
+func (node *quotaNode) String() string {
+	return fmt.Sprintf("qName:%s, req:%v, SharedWeight:%v, minQuota:%v, runtimeQuota:%v",
+		node.quotaName, node.request, node.SharedWeight, node.minQuota, node.runtimeQuota)
+}
+
+func NewQuotaNode(quotaName QuotaNameType, SharedWeight, request, min int64, allowLentResource bool) *quotaNode {
+	if SharedWeight == 0 {
+		return nil
+	}
+
+	return &quotaNode{
+		quotaName:         quotaName,
+		request:           request,
+		SharedWeight:      SharedWeight,
+		minQuota:          min,
+		runtimeQuota:      0,
+		allowLentResource: allowLentResource,
+	}
+}
+
+type quotaTree struct {
+	quotaNodes map[QuotaNameType]*quotaNode
+}
+
+func NewQuotaTree() *quotaTree {
+	return &quotaTree{
+		quotaNodes: make(map[QuotaNameType]*quotaNode),
+	}
+}
+
+func (qt *quotaTree) insert(groupName QuotaNameType, SharedWeight, request, min int64, allowLentResource bool) {
+	if _, exist := qt.quotaNodes[groupName]; !exist {
+		qt.quotaNodes[groupName] = NewQuotaNode(groupName, SharedWeight, request, min, allowLentResource)
+	}
+}
+
+func (qt *quotaTree) erase(groupName QuotaNameType) {
+	if _, exist := qt.quotaNodes[groupName]; exist {
+		delete(qt.quotaNodes, groupName)
+	}
+}
+
+func (qt *quotaTree) update(groupName QuotaNameType, SharedWeight, request, min int64, allowLentResource bool) {
+	if nodeValue, exist := qt.quotaNodes[groupName]; exist {
+		if nodeValue.SharedWeight != SharedWeight || nodeValue.request != request || nodeValue.minQuota != min ||
+			nodeValue.allowLentResource != allowLentResource {
+			qt.quotaNodes[groupName] = NewQuotaNode(groupName, SharedWeight, request, min, allowLentResource)
+		}
+	}
+}
+
+func (qt *quotaTree) find(groupName QuotaNameType) (bool, *quotaNode) {
+	if nodeValue, exist := qt.quotaNodes[groupName]; exist {
+		return exist, nodeValue
+	}
+
+	return false, nil
+}
+
+func (qt *quotaTree) adjustQuota(totalResource int64) {
+	toPartitionResource := totalResource
+	totalSharedWeight := int64(0)
+	needAdjustQuotaNodes := make([]*quotaNode, 0)
+	for _, node := range qt.quotaNodes {
+		runtimeQuotaTmp := node.request
+		if runtimeQuotaTmp > node.minQuota {
```

Comment:
if runtimeQuotaTmp > node.minQuota  -> if node.request > node.minQuota;  althrough the effect is same, but request is much more reasonable